File: lcs/lcs.cpp
// longest common sequences to find the gene sequence
//
#include <iostream>
#include <vector>
#include <string>

using namespace std;

int lcs(const string& s1, const string& s2) {
    int m = s1.size()+1;
    int n = s2.size()+1;
    vector<vector<int>> opt(m, vector<int>(n, 0));
    for(int i=1; i < m; ++i) {
       for(int j=1; j < n; ++j) {
           if(s1[i-1]==s2[j-1])
               opt[i][j]=1+opt[i-1][j-1];
           else
               opt[i][j]=max(opt[i-1][j], opt[i][j-1]);
       }
    }
    return opt[m-1][n-1];
}

int main() {
    string s1="gtactcatg";
    string s2="agtcatag";
    cout << s1 << endl;
    cout << s2 << endl;
    cout << lcs(s1, s2) << endl;
}

File: lcs/lcs.py
#!/usr/bin/python3
#// longest common sequences to find the gene sequence
#//


def lcs(s1, s2):
    m = len(s1)+1
    n = len(s2)+1
    opt = [[0]*n for i in range(m)]
    for i in range(1, m):
        for j in range(1, n):
            if(s1[i-1]==s2[j-1]):
                opt[i][j]=1+opt[i-1][j-1]
            else:
                opt[i][j]=max(opt[i-1][j], opt[i][j-1])

    return opt[m-1][n-1];

def main(): 
    s1="gtactcatg";
    s2="agtcatag";
    print(s1)
    print(s2)
    print(lcs(s1, s2))

if __name__ == "__main__":
    main()

